op-batcher/batcher/channel_manager.go

```diff
@@ -301,15 +304,21 @@ func (s *channelManager) registerL1Block(l1Head eth.BlockID) {
 func (s *channelManager) processBlocks() error {
 	var (
 		blocksAdded int
-		_chFullErr  *ChannelFullError // throw away, just for type checking
 		latestL2ref eth.L2BlockRef
 	)
-	for i, block := range s.blocks {
+	for i := 0; ; i++ {
+		block, ok := s.blocks.Dequeue()
```

Comment:
Maybe we add a `Queue.Peek() (T, bool)` function that returns the next queued element without dequeueing it, so we don't have to enqueue again in case of errors? We would then call `Dequeue` after the error checks, throwing away its return values.

The current refactor is less efficient in terms of memory management compared to the old code.

Reply:
Good idea [2a1a63d](https://github.com/ethereum-optimism/optimism/pull/12180/commits/2a1a63dd5020caf7abe4e8e4a5395ef741bfad96)

Reply:
In order to be able to be even more memory efficient and retain underlying slice operations similar to how it was before, we could also add `PeekN(int)` and `DequeueN(int)` functions, so we only read the queue elements one by one, then dequeue all at once.

Reply:
[use queue.PeekN and queue.DequeueN](https://github.com/ethereum-optimism/optimism/pull/12180/commits/6c2c349a17a1c427f9f5212c1235269375a33d5c)